src/System.Devices.Gpio/GPIORef.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Devices.Gpio
+{
+    /// <summary>
+    /// GPIO - Basic Scenarios
+    /// 
+    /// Most GPIO implementations share a core set of functionality to allow basic on/off control
+    /// of pins. At the very least, we should support these.
+    /// - Open a object that is a representation of a pin with the given pin number
+    /// - Support closing a pin to release the resources owned by that pin object
+    /// - Represent the mode of the pin that details how the pin handles reads and writes (e.g. Input, Output)
+    /// - Allow a resistor to be added to a pin such that it can be set as pullup or pulldown (or no pull)
+    /// - Support setting a PWM value on a pin
+    /// </summary>
+    #region GPIO - Basic Scenarios
+
+    public partial class GPIOPin : IDisposable
+    {
+        int PinNumber { get { throw new NotImplementedException(); } }
+        bool Read() { throw new NotImplementedException(); }
+        void Write(bool value) { throw new NotImplementedException(); }
+        void Dispose() { throw new NotImplementedException(); }
+        GPIOPin(GPIOController controller, int pinNumber, GPIOPinMode pinMode = GPIOPinMode.Input) { }
+
+        GPIOPinMode PinMode { get { throw new NotImplementedException(); } set { throw new NotImplementedException(); } }
+        bool IsPinModeSupported(GPIOPinMode pinMode) { throw new NotImplementedException(); }
+        int PWMValue { set { throw new NotImplementedException(); } }
+    }
+
+    public partial class GPIOController : IDisposable
+    {
+        GPIOPin this[int pinNumber] { get { throw new NotImplementedException(); } }
```

Comment:
What's the difference between getting a pin by opening it or by using the indexer?